### ISSUE TITLE
Redirect should encode URLs if it needs to

### ISSUE THREAD

**Issue by miloush** [aspnet/Mvc#8518]
All these issues are result of people trying to redirect to a URL that is not all ASCII. Resolution seems to be "you should have encoded your URL".

https://github.com/aspnet/Home/issues/2678
https://github.com/aspnet/Mvc/issues/7701
https://github.com/aspnet/KestrelHttpServer/issues/2508
https://github.com/aspnet/KestrelHttpServer/issues/2296
https://github.com/aspnet/Security/issues/1646
https://github.com/aspnet/HttpAbstractions/issues/971
https://github.com/aspnet/Routing/issues/513
https://github.com/aspnet/Mvc/issues/7529
https://github.com/aspnet/HttpAbstractions/issues/907
https://github.com/aspnet/Mvc/issues/6609
https://github.com/aspnet/Home/issues/2106
https://github.com/madskristensen/Miniblog.Core/issues/87

However, the fact that the RedirectResult puts the URL into a HTTP header (which is the ultimate source of the issue) is an implementation detail that the caller cannot magically assume. It is therefore responsibility of e.g. RedirectResultExecutor to meet the requirements and ensure the URL is properly formatted in case it decides to use HTTP headers for redirection.

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us, @miloush.
@Tratcher, what are your thoughts regarding this?

**Comment by Tratcher** [comment]
We have helpers for generating properly encoded urls like [UriHelper.BuildAbsolute](https://github.com/aspnet/HttpAbstractions/blob/e8d5a85c9949e26914abc8b96594f741725c0442/src/Microsoft.AspNetCore.Http.Extensions/UriHelper.cs#L48). By the time you have combine everything into one string it's too late automatically encode, some parts are ambiguous like ? and # if present in the path.

Perhaps overhauling RedirectResult to take in PathString, QueryString, etc. rather than a full string?

**Comment by Tratcher** [comment]
Right, we need to be specific about the kind of encodings. To create an unambiguous url there are several syntax characters you may need to escape like `/?#:%.&` etc. since they affected the meaning of the individual parts. Then you may need to encode it for transport / embedding in your specific, especially unicode, but # and other symbols can also be problematic in certain context like HTML.

PathString and company are specifically designed to encode URLs for use in HTTP headers. They take care of both the syntax escaping when you combine components, as well as fully encoding any unicode. It's important to do both of these at the same time so you don't double encode anything.

Unicode has a long messy relationship with HTTP headers and it's best to be explicit about the encoding.

**Comment by Tratcher** [comment]
I've filed https://github.com/aspnet/Home/issues/3574 for a Response.Redirect overload that accepts the components.

**Comment by miloush** [comment]
Okay so nowadays when one wants to fix this issue, they need to create a special ASCII encoded URL to pass to the Redirect method, and use the Unicode URL in links and on pages, e.g. https://github.com/madskristensen/Miniblog.Core/commit/14ec33b42fcd1d9e5f44ceff85749189a614d8dd

Imagine you add the suggested overload to Response.Redirect that takes the components, how would that eliminate the need for such fix?

**Comment by Tratcher** [comment]
[UriHelper.GetDisplayUrl](https://github.com/aspnet/HttpAbstractions/blob/e8d5a85c9949e26914abc8b96594f741725c0442/src/Microsoft.AspNetCore.Http.Extensions/UriHelper.cs#L196) is what I refer people to when they want the friendly display for a request url. The signature doesn't work well for self-generated urls (it takes an HttpRequest), but that's another case where adding an overload that takes in the same components and renders them in a user displayable format would make sense. Note that one doesn't do the syntax escaping, you'd only use it for the display text of a link and use the encoded version for the actual link.

**Comment by miloush** [comment]
Okay so for self-generated urls, you would recommend a sort of GetLink() method that returns `(scheme, HostString, PathString, QueryString, FragmentString)` and then feeding it to either Redirect or GetDisplayUrl? I am worried that just sticking to a string instead is much more convenient...

**Comment by Tratcher** [comment]
Understood on the convenience of a string. The trouble is that it doesn't provide any context, the user has to know if it's encoded or decoded and handle it accordingly. I'm starting to think we need a url structure that aggregates all of the components for easy flow. (Not System.Uri because that causes quite a few transformations and doesn't encode appropriately for HTTP headers or HTML pages).

**Comment by Tratcher** [comment]
@mkArtakMSFT Looking at RedirectResult/Executor it's a 1:1 mapping to Response.Redirect. @miloush is certainly right on one main point, these API do not tell you what to do nor help you do it successfully. The doc comments make no reference to any encoding requirements. I will dispute the bit about putting the value into a HTTP header being an implementation detail, that's the core mechanic for HTTP redirects, this API is only sugar around that. Unfortunately it's not enough sugar. Changing the API to take in PathString, QueryString, etc. or a higher level construct would give users much more guidance here.

**Comment by mkArtakMSFT** [comment]
Thanks @Tratcher.
@javiercn, can you please handle the proposed change? Thanks!

**Comment by javiercn** [comment]
I understand the concerns here. As far as I understand this is just a limitation for the location header, on other places (other than HTML encoding) is totally valid to have non ASCII characters on the Uri. 

That said, I'm not sure adding new APIs will do anything to solve this problem. Most people will simply keep calling the string based API (as is the most convenient one) to later realize that the Url is not properly encoded. 

When that happens, I don't really see an advantage in adding one overload for each of the redirect methods that takes all the individual components as HostString, PathString, etc compared to using UriHelper before calling on the redirect methods as suggested above.

I think the main thing we can change here is to make generating correctly encoded urls in MVC easier. This is something that we can add to UrlHelper. That way, we don't needs these overloads and this just becomes or something similar.
```csharp
return Redirect(UrlHelper.CreateUrl(host,port,base,path,query,fragment));
```

**Comment by Tratcher** [comment]
Efficient. It's still lacking in discoverability though.

**Comment by miloush** [comment]
I am not going to object any new overloads but I am with Javier that I don't think it will do much to solve the problem. "I'm starting to think we need a url structure" sounded like closest to what would help.

But I understand and value the framework is built on simple types and it is predictable what gets output.

My proposal was that instead of ValidateHeaderCharacters throwing, it would ASCII encode the header value. Why would that be a bad idea? You get zero performance loss in already conforming cases since the check needs to be done anyway.

**Comment by rynowak** [comment]
This would double-encode someone's already-endcoded string

**Comment by miloush** [comment]
> This would double-encode someone's already-endcoded string

No. If somebody already encoded the string it would be all ASCII and nothing would get changed on the way.

**Comment by Tratcher** [comment]
Encoding deals with more than non-ascii characters. You can't really separate the encoding of syntax characters from the encoding of unicode characters, both need to be done at the same time.

**Comment by javiercn** [comment]
As @rynowak mentioned.
> This would double-encode someone's already-endcoded string

Having RedirectResult do this for you would be a breaking change, so I don't think we want to go that way.

We should just make this simpler by adding a convenience API on UrlHelper (The place used to generate Urls in MVC) and maybe if possible an analyzer to detect simple bad patterns. (Like constructing the url on the argument to redirect through + or as a string literal,  or in a variable)

**Comment by miloush** [comment]
As I said that is not true, unless I am missing something obvious. Could you give me an example where this would break something?

**Comment by javiercn** [comment]
@miloush It changes the response in an observable way. That has the potential to break clients relying on the current behavior.

This can be done, but it would require at least to be opt-in and require a compatibility flag to prevent breaking people upgrading.

**Comment by miloush** [comment]
The current behavior is to throw an exception - that is the only case when there would be a difference, all current cases which result in an actual response would stay intact. Does that count as changing response in an observable way and worth compatibility flag?

I am not religious about doing this, but people keep running into the problem and this seems to be the right place to fix it to make everyone's life easier.
